Allocate domain state using malloc instead of mmap

abbysmal opened the pull request:
This pull request addresses ocaml-multicore/ocaml-multicore#796. 

The motivation is to avoid unnecessary use of `mmap` in the runtime. The PR replaces the use of `mmap`, which is a vestige of the old multicore GC design, with `malloc` (through `caml_stat_alloc_noexc`). 

The dynamically allocated domain state structures are not freed similar to the current `mmap` based implementation. This is because the runtime stores the GC stats in the domain state structures, and freeing them at domain termination would lose stats. The proper solution to this would be to orphan the stats to a global structure in the runtime. At this point, we will also get the benefit of being able to provide accurate GC stats per domain (for which no API exists now). This should be addressed in a separate PR.

There is also no attempt to align the domain state structure to cache lines as we suggested earlier in ocaml-multicore/ocaml-multicore#796 discussion. This needs experimental evidence for the purported benefits.

cc @kayceesrk @ctk21.

kayceesrk committed:
Use malloc for domain state instead of mmap

xavierleroy commented:
By "statically" you mean "using `malloc`", right?  For old C programmers, "statically" means allocated once and for all in global variables or "static"-qualified local variables, and malloc-ed memory is "dynamically" allocated.

abbysmal commented:
@xavierleroy Indeed, this is the wrong wording, I'll edit that after investigating the testsuite crash. :)

xavierleroy commented:
Note that you could have a global array of `Max_domains` elements of type `caml_domain_state`...  and then it would really be static allocation :-)  But I don't see any benefit to this static allocation.

abbysmal commented:
Indeed, I think it is fine that way.
I'm still currently investigating this issue, could not get it to trigger on any of the machines I could get my hands on, I'm currently in the process of excavating a coredump from the GH runs, which can reliably fail.
I will report back once I made some progresses.

abbysmal commented:
@kayceesrk found the fix to the issue showing up on the CI, it is required to properly zero out the allocated memory.

Title and original description were also edited.

abbysmal commented:
@kayceesrk 
https://github.com/ocaml/ocaml/runs/4965800849?check_suite_focus=true#step:5:4962 is actually quite strange, I am not very familiar with this testcase (I think @ctk21 took a look at it a while ago?).

https://github.com/ocaml/ocaml/pull/10950/commits/862cc0f7fc10b6c2f6b97c575e13f1973eaf3b20 is perhaps not a good idea and there may be something bad happening there, where `caml_startup_common` seems to be disappearing from the stack walking code result using libunwind?

abbysmal commented:
Actually this is not related to this branch and has been triggered before on another PR.
https://github.com/ocaml/ocaml/runs/4966505783?check_suite_focus=true
I will revert this change, this is not this PR's job to address it. I'll open a follow-up issue to track this specific failure.

kayceesrk commented:
LGTM. I developed this PR along with @Engil. Hence, I would prefer it if someone else has a look and approves the changes.

abbysmal commented:
And I can see that https://github.com/ocaml/ocaml/commit/acc490f8002ed53237a22727c761845226d0086a addressed the `unwind/driver.ml` failure yesterday, my bad. :-)

xavierleroy commented:
> And I can see that acc490f addressed the unwind/driver.ml failure yesterday, my bad. :-)

Nothing to feel bad about, and I hope the commit message explains why the backtrace changed (a call became a tail call).

avsm commented on runtime/domain.c:
```diff
@@ -392,49 +389,51 @@ int caml_reallocate_minor_heap(asize_t wsize)
+    if (!d->state) {
+      /* FIXME: Never freed. Not clear when to. */
+      domain_state =
+        (caml_domain_state*)caml_stat_alloc_noexc(sizeof(caml_domain_state));
```
You can use `caml_stat_calloc_noexc` here and get rid of the memset below.

abbysmal replied:
Thanks!

abbysmal commented:
I made a few changes and I believe the PR is now good for review again.

avsm commented on Changes:
```diff
@@ -17,6 +17,11 @@ Working version
   (David Allsopp, review by Xavier Leroy and Damien Doligez)
 
+- #10950: Do not use mmap to allocate Caml_state.
+  In order to reduce virtual memory usage, we statically allocate
```
I think you're dynamically allocating the structure now..

abbysmal replied:
Right, good catch.

avsm commented on runtime/domain.c:
```diff
@@ -235,7 +232,7 @@ static dom_internal* next_free_domain() {
 
 #ifdef __APPLE__
 /* OSX has issues with dynamic loading + exported TLS.
-    This is slower but works */
```
Unrelated to this PR, but: it would be good to check if this macOS workaround is still needed.  Reading up on [a recent TLS implementation survey](https://maskray.me/blog/2021-02-14-all-about-thread-local-storage) seems to indicate that the situation has improved there (for `thread_local` anyway).

abbysmal replied:
Indeed, should we open an issue to keep track of this?

avsm commented:
This looks right to me, but it does interfere dramatically with https://github.com/ocaml/ocaml/pull/10964. Who wins? :-)

abbysmal commented:
@avsm (continuing from an off-site conversation.) This PR does not actually change the handling of Max_domains so I think this does not quite interfere, or am I missing a detail?

Another review on this would be welcome (maybe @xavierleroy ?)

kayceesrk reviewed (approved):
LGTM!

FYI I wrote this PR along with @Engil.

gasche commented on runtime/domain.c:
```diff
@@ -417,50 +414,52 @@ static uintnat fresh_domain_unique_id(void) {
+    if (!d->state) {
+      /* FIXME: Never freed. Not clear when to. */
+      domain_state = (caml_domain_state*)
+        caml_stat_calloc_noexc(1, sizeof(caml_domain_state));
```
If I understand correctly, this is not actually a leak, at least in "being nice with shutdown mode" because `caml_stat_calloc_noexc` is pooled and will be properly reclaimed by `caml_shutdown`.

Why don't we have a `caml_free_domains` function that would be de-allocating parent of `caml_init_domains`, and would be called by `caml_shutdown`? Is this simply an omission, something that would be useless, or an actively bad idea for some reason I am missing?

gasche commented on runtime/domain.c:
```diff
@@ -417,50 +414,52 @@ static uintnat fresh_domain_unique_id(void) {
+      /* FIXME: Never freed. Not clear when to. */
+      domain_state = (caml_domain_state*)
+        caml_stat_calloc_noexc(1, sizeof(caml_domain_state));
+      if (domain_state == NULL)
```
Minor note: this code uses `if (!d->state)` but also `if (domain_state == NULL)`: implicit-NULL-as-false and explicit-NULL-comparison. I think that the runtime house style is rather the latter, so `if (d->state == NULL)` would be more familiar to me.

(It's fine to leave the code as-is, but if you do make a change, maybe we could flip the test to check the `!= NULL` case first and have the short already-initialized branch out of the way?)

gasche commented on runtime/domain.c:
```diff
@@ -417,50 +414,52 @@ static uintnat fresh_domain_unique_id(void) {
+    caml_plat_wait(&all_domains_cond);
 
   d = next_free_domain();
   if (d) {
```
Nitpick: It's rather hard to find the end of this long branch. Could we maybe use `if (d == NULL) goto init_domain_complete` as inside the branch, and not have a super-long conditional block?

gasche commented on runtime/domain.c:
```diff
@@ -417,50 +414,52 @@ static uintnat fresh_domain_unique_id(void) {
-    if (!s->interrupt_word) {
-      caml_domain_state* domain_state;
-      atomic_uintnat* young_limit;
-      /* never been started before, so set up minor heap */
```
Minor note: this comment was wrong in the first place I think, as the minor heap is not set here. (I noticed this while reading the code with @sabine this morning.)

gasche commented on runtime/domain.c:
```diff
@@ -417,50 +414,52 @@ static uintnat fresh_domain_unique_id(void) {
-      domain_state = (caml_domain_state*)(d->tls_area);
-      young_limit = &domain_state->young_limit;
-      s->interrupt_word = young_limit;
-      atomic_store_rel(young_limit, (uintnat)domain_state->young_start);
```
What happened to the initialization of `young_limit` in this patch, is it missing from the new code?

abbysmal replied:
@gasche this was a redundant initialization, `young_limit` is properly set when allocating the minor heap with `caml_reallocate_minor_heap`.

gasche reviewed (commented):
I looked at the patch. Generally I like that it is simpler.

Unless I'm missing something, the new code does not initialize `domain_state->young_limit` anymore, and I wonder if this is a bug or the initialization was redundant/useless for some reason.

Some lines of the patch are a bit confusing initially because they replace `d->tls_area` with `d->state`. Unless I'm missing something, it was an invariant from the previous implementation that, on properly-initialized domains, those two pointer fields are always equal. (One shows an untyped view of the underlying memory, the other presents it as a domain_state structure.) It was a bit confusing, but it's gone now.

abbysmal committed:
runtime: streamline NULL checking on domain_state in create_domain

gasche reviewed (approved):
Thanks for the clarification, this looks good.

abbysmal committed:
runtime: in create_domain, do not hide failure to get a domain behind a condition.

Instead we just check for d == NULL and goto terminate label.

abbysmal commented:
@gasche I pushed some changes as you suggested, thank you for your review.

[c4312b2](https://github.com/ocaml/ocaml/pull/10950/commits/c4312b27730cb100ac75f96b3b17302f008d9223) is especially noisy because I had to reindent the whole block.
I call pull it out of this PR if you feel like it is better this way.

I rebased and squashed other intermittent changes. (sorry for the noise.)

As for the `domains_free` function, I think this may be a worthy addition, and from the top of my head I'm not sure there's something strictly forbidding this, I think it outgrows a bit this PR, but we should maybe look into this.

gasche commented:
Using the "Hide whitespace" option of github makes it obvious what the changes does: [https://github.com/ocaml/ocaml/pull/10950/commits/8351a7553ba701f2baebd39eab3478f610d68aea?diff=split&w=1](https://github.com/ocaml/ocaml/pull/10950/commits/8351a7553ba701f2baebd39eab3478f610d68aea?diff=split&w=1).

> As for the domains_free function, I think this may be a worthy addition, and from the top of my head I'm not sure there's something strictly forbidding this, I think it outgrows a bit this PR, but we should maybe look into this.

Certainly, this was an idea for later, not something that should be in this PR.

@Engil Everything is green now. Would you like to click the "merge" button yourself?

abbysmal commented:
@gasche sure thing, thank you. :)
